UC20 spike changes

cmatsuoka opened the pull request:
This is the set of UC20 spike changes up to London names and extracted kernel for encrypted ubuntu-data.

mvo5 committed:
gadget.yaml: add system-recovery partition

For UC20 we want to have a recovery partition. This adds to
the existing system-boot partition. The size is currently set
to 600Mb - this is a bit arbitrary and will allow us roughtly
2 recovery systems. We have the following sizes today:
 * core18:  54MB
 * snapd:   23MB
 * gadget:   1MB
 * kernel: 200MB
The recovery partition is setup after the system-boot
partition.

It also adds a initial grub.cfg which is very basic and
only supports a single recovery system at this point but
that recovery system can be selected via the grubenv:
- `snap_recovery_kernel`
- `snap_recovery_core`
settings.

mvo5 committed:
recovery.md: add recovery partition overview.

mvo5 committed:
improve recovery boot menu

mvo5 committed:
fix chainloading

mvo5 committed:
use static recovery grub entry

mvo5 committed:
fix typos in recovery.md

mvo5 committed:
gadget.yaml: increase size of recovery to 1G

mvo5 committed:
gadget.yaml: increase size system-boot to 100Mb (for unpacked kernels)

mvo5 committed:
drop use of "regexp" module

mvo5 committed:
add new "role: system-recovery"

mvo5 committed:
clarify how first-boot works

mvo5 committed:
update for latest changes

Update for latest changes in
  https://github.com/CanonicalLtd/ubuntu-image/pull/171
and
  https://github.com/snapcore/snapd/pull/6899

mvo5 committed:
Merge remote-tracking branch 'upstream/20' into 20

cmatsuoka committed:
handle recover_reboot mode and add debug shell

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

mvo5 committed:
Merge pull request #1 from cmatsuoka/writable-ramdisk

handle recover_reboot mode and add debug shell

mvo5 committed:
snapcraft: move to base: core20

mvo5 committed:
grub.cfg-recovery: update for new /systems/snaps/$snap_name layout

In London we decided to use a new directory layout and put the
snaps into /snaps on the ubuntu-seed (recovery) partition. This
PR updates the grub config to use the right kernel.

mvo5 committed:
many: use lodon sprint partiton names

During the London sprint we decided to use the following partition
names:
- ubuntu-seed (previously: sys-recover)
- ubuntu-boot (previously: system-boot)
- ubuntu-data (previously: writable)

This commit updates the code to use these names.

cmatsuoka committed:
grub.cfg-normal: boot from extracted kernel

To use an encrypted writable data partition, we must boot from kernels
extracted to the ubuntu-boot partition.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

vorlonofportland commented on grub.cfg-normal:
```diff
@@ -31,12 +31,11 @@ elif [ "$snap_mode" = "trying" ]; then
-    linux (loop)/kernel.img $cmdline
-    initrd (loop)/initrd.img
+    search --label ubuntu-boot --set=ubuntu_boot
+    linux ($ubuntu_boot)/kernel.img $cmdline systemd.debug-shell=1
```
'systemd.debug-shell=1' 
This looks like an unrelated change that doesn't belong on the actual core20 branch.

mvo5 replied:
This is there to make debugging easier, if you feel strongly about it we could remove it. Note that this is the "uc20" spike branch we are merging into - this is not a "production" branch or anything like this.

vorlonofportland replied:
How will we make sure we clean up all debugging-only code when we go to production?  Spike or not, that's what worries me since this doesn't interfere with the booting of the system but does leave a security hole around.

cmatsuoka replied:
As agreed before starting the spike, this branch won't be merged into master -- it will serve only as a reference and all changes will be properly rewritten at a later moment, which should serve as a filter to remove debugging code. That said, most early problems related to snapd seeding have already been fixed so I believe we don't need the debug shell anymore. I'll push a change to remove it.

vorlonofportland commented on gadget.yaml:
```diff
@@ -14,15 +14,29 @@ volumes:
-        filesystem-label: system-boot
-        size: 50M
+        filesystem-label: ubuntu-boot
+        size: 100M
```
I'm skeptical of this as a size for the partition given that we will need to store 2x vmlinuz (+ embedded initramfs) on this fs for the FDE case.

mvo5 replied:
What size do you have in mind? I just checked and the kernel is right now around 8Mb for vmlinuz and around 7Mb for the initamfs (this includes some basic tools to unlock the LUKS device).

vorlonofportland replied:
initramfs will include all of the microcode updates, which are currently ~2.5M; one golang binary (snapd is currently 17M on amd64); systemd (~2M or so?); C runtime (at least 14M, based on the library dependencies of systemd+udev).

The microcode doesn't get compressed.  The other bits may get compressed, but there's an ongoing discussion whether compressing them negatively impacts boot speed, so we shouldn't assume they are compressed (or compressed much).

That's 35.5M for the initramfs, not counting any kernel modules which might need to be shipped in the embedded initramfs instead of built in to the kernel image; plus 8.4M for vmlinuz, which puts us at ~44M.

If we expect these systems to be usable for 10 years, we'll need to expect some expansion of kernels, if only due to the addition of drivers for new hardware.

So 150 or 200M seems more suitable to me (rounded to 4MiB).

xnox replied:
30M	ubuntu-core-initramfs.img
9.0M /boot/vmlinuz-5.2.0-8-generic
is what i have currently, without firmware, microcode, snapd. But this has too many kernel modules in the initrd.

vorlonofportland commented on Makefile:
```diff
@@ -69,4 +69,4 @@ all:
 install:
 	install -m 644 pc-boot.img pc-core.img shim.efi.signed grubx64.efi $(DESTDIR)/
-	install -m 644 grub.conf grub.cfg $(DESTDIR)/
+	install -m 644 grub.cfg-recovery grub.cfg-normal grub.conf $(DESTDIR)/
```
'grub.conf' is not a native name for the config file on grub's side, this name exists for compatibility with snapd's requirements. If there are now two grub.cfg's, should the grub.conf symlink be dropped?

mvo5 replied:
I think we can drop the symlink now - I'm not sure about the consequences though - I would like to do that in a followup so that if things break we don't need to look at too many variables.

vorlonofportland replied:
+1 to deferring.

vorlonofportland commented on grub.cfg-recovery:
```diff
@@ -0,0 +1,40 @@
+# FIXME: find a better way? this ensures that someone instructed snapd
+#        to go into "install" mode it actually does that
+if [ "$snap_mode" = "install" ]; then
+    set default=1
```
Discussed in London that these don't need to be menu entries at all: if the chooser interaction is expected to be entirely via the initramfs, this could execute the boot commands inline instead of driving a menu selection.

mvo5 replied:
Yes, mostly for convenience today to see things flipping from "recovery" -> "normal" (a bit of a debug aid).

cmatsuoka committed:
many: remove kernel and core command line parameters

Read snap_core and snap_kernel directly from the grub environment
file instead of passing them as kernel command line parameters.
Also hide the bootloader menu and disable the debug shell.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

cmatsuoka committed:
recovery.md: update partition names

Use London names (ubuntu-seed, ubuntu-boot, etc) for disk partitions.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

xnox commented:
This looks good, impossible to build with core20 based atm, will switch to core18 base for the time being.
